keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1191 @@
+to run to completion.
+
+As a result, the current implementation does not cover the situation where the workload
+is truly embarrassingly parallel and each index is independent of other indexes.
+
+For instance, if indexed jobs were used as the basis for a suite of long-running integration tests,
+then each test run would only be able to find a single test failure.
+
+Other popular batch services like AWS Batch use a separate backoff limit for each index,
+showing that this is a common use case that should be supported by Kubernetes.
+
+### Goals
+
+- allow to count failures towards the backoffLimit independently for all indexes,
+- allow to continue Job execution despite some of its indexes failing,
+- allow to fail an index (stop recreating pods for the index) using pod failure policy.
+
+### Non-Goals
+
+- allow to control the number of retries per index when pod's `restartPolicy=OnFailure`
+(see [Support backoffLimitPerIndex when restartPolicy=OnFailure](#support-backofflimitperindex-when-restartpolicyonfailure)).
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+We propose a new policy of running Indexed Jobs in which the backoff limit
+controls the number of retries per index. When the new policy is used all
+indexes execute till their success or failure. However, we also propose a new
+API field to control the number of failed indexes.
+
+Additionally. we propose a new action in PodFailurePolicy, called FailIndex,
+to short-circuit failing of the index before the backoff limit per index is
+reached.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a CI/CD platform administrator, I want to use Indexed Jobs to run
+suites of integration tests, one suite per index. A failure of one suite should
+not interrupt running of other suites. Additionally, I would like to be able
+to control the maximal number of retries per index.
+
+The following Job configuration could satisfy my use case:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimit: 1
+  runPolicy:
+    backoffLimitPerIndex: true
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+```
+
+In this case, we run 10 indexes representing the test suites. We allow for one
+failure per index.
+
+#### Story 2
+
+As a CI/CD platform administrator from the [Story 1](#story-1) I want to be able
+to control the failures with the pod failure policy. In particular, I want
+to be able to use pod failure policy to avoid restarts of some indexes, based
+on exit codes.
+
+The following Job configuration could satisfy my use case:
+
+```yaml
+apiVersion: v1
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  backoffLimit: 1
+  runPolicy:
+    backoffLimitPerIndex: true
```

Comment:
The default is 6 as the default in the regular case. The rationale is to avoid confusion of the users with conditional defaults. I have added a section in alternatives on this. I'm open to reassess this based on the community feedback.

Reply:
Resolving as the current proposal is to have `backoffLimitPerIndex` which is an int32 without a default.